File: Problems/Problem_3/Problem_3.cpp
#include<iostream>

using namespace std;

bool isPrime(long long int N){
    //2 is the only even prime
    if(N==2 || N==3){
        return true;
    }
    //Non-neighbours of 6 except 2 and 3 are non-primes
    else if(N%6!=1 && N%6!=5){
        return false;
    }
    //we only need to check if any odd number less
    //than the square root of N divides N
    for(long long int i=3; i*i <= N; i+=2){
        if(N % i == 0){
            return false;
        }
    }
    return true;
}

int main(){
    long long int N,i;
    cin>>N;
    for(i=2; N>0; i++) {
        if(isPrime(i)) {
            N--;
        }
    }
    cout<<i-1;
    return 0;
}

File: Problems/Problem_3/Problem_3.py
def isPrime(n):
    if n == 2:
        return True

    # Even numbers above 2 can't be prime
    if n % 2 == 0:
        return False

    # All primes above 3 are neighbors of a multiple of 6
    if n>3 and (n+1) % 6 != 0 and (n-1) % 6 != 0:
        return False

    # For the rest, we check up to sqrt(n) for factors
    i = 3
    while i*i <= n:
        if n % i == 0:
            return False
        i+=2

    return True

def main():
    n = int(input())
    p = 2
    for i in range(n,0,-1):
        while not isPrime(p):
            p+=1
        p+=1
    print(str(n)+"th prime: "+str(p-1))

main()
